pkg/controllers/management/gke/gke_cluster_handler.go

```diff
@@ -0,0 +1,519 @@
+					}
+					return cluster, statusErr
+				}
+				cluster, statusErr = e.SetFalse(cluster, apimgmtv3.ClusterConditionWaiting,
+					fmt.Sprintf("failed to communicate with cluster: %v", err))
+				if statusErr != nil {
+					return cluster, statusErr
+				}
+				return cluster, err
+			}
+		}
+
+		cluster, err = e.recordAppliedSpec(cluster)
+		if err != nil {
+			return cluster, err
+		}
+		return e.SetTrue(cluster, apimgmtv3.ClusterConditionUpdated, "")
+	case "updating":
+		cluster, err = e.SetTrue(cluster, apimgmtv3.ClusterConditionProvisioned, "")
+		if err != nil {
+			return cluster, err
+		}
+
+		e.ClusterEnqueueAfter(cluster.Name, enqueueTime)
+		if failureMessage == "" {
+			logrus.Infof("waiting for cluster GKE [%s] to update", cluster.Name)
+			return e.SetUnknown(cluster, apimgmtv3.ClusterConditionUpdated, "")
+		}
+		logrus.Infof("waiting for cluster GKE [%s] update failure to be resolved", cluster.Name)
+		return e.SetFalse(cluster, apimgmtv3.ClusterConditionUpdated, failureMessage)
+	default:
+		if cluster.Spec.GKEConfig.Imported {
+			cluster, err = e.SetUnknown(cluster, apimgmtv3.ClusterConditionPending, "")
+			if err != nil {
+				return cluster, err
+			}
+			logrus.Infof("waiting for cluster import [%s] to start", cluster.Name)
+		} else {
+			logrus.Infof("waiting for cluster create [%s] to start", cluster.Name)
+		}
+
+		e.ClusterEnqueueAfter(cluster.Name, enqueueTime)
+		if failureMessage == "" {
+			if cluster.Spec.GKEConfig.Imported {
+				cluster, err = e.SetUnknown(cluster, apimgmtv3.ClusterConditionPending, "")
+				if err != nil {
+					return cluster, err
+				}
+				logrus.Infof("waiting for cluster import [%s] to start", cluster.Name)
+			} else {
+				logrus.Infof("waiting for cluster create [%s] to start", cluster.Name)
+			}
+			return e.SetUnknown(cluster, apimgmtv3.ClusterConditionProvisioned, "")
+		}
+		logrus.Infof("waiting for cluster GKE [%s] pre-create failure to be resolved", cluster.Name)
+		return e.SetFalse(cluster, apimgmtv3.ClusterConditionProvisioned, failureMessage)
+	}
+}
+
+// setInitialUpstreamSpec builds a view of the upstream cluster and adds it to the status of the cluster resource
+func (e *gkeOperatorController) setInitialUpstreamSpec(cluster *mgmtv3.Cluster) (*mgmtv3.Cluster, error) {
+	logrus.Infof("setting initial upstreamSpec on cluster [%s]", cluster.Name)
+	cluster = cluster.DeepCopy()
+	upstreamSpec, err := clusterupstreamrefresher.BuildGKEUpstreamSpec(e.SecretsCache, cluster)
+	if err != nil {
+		return cluster, err
+	}
+	cluster.Status.GKEStatus.UpstreamSpec = upstreamSpec
+	return e.ClusterClient.Update(cluster)
+}
+
+// updateGKEClusterConfig updates the GKEClusterConfig object's spec with the cluster's GKEConfig if they are not equal.
+func (e *gkeOperatorController) updateGKEClusterConfig(cluster *mgmtv3.Cluster, gkeClusterConfigDynamic *unstructured.Unstructured, spec map[string]interface{}) (*mgmtv3.Cluster, error) {
```

Comment:
It looks like this function could have a more generic name and moved to `clusteroperator/utils.go`. This would also require the EKS version to be deleted in favor of the more generic version.

Reply:
This one is a little tricky because it invokes `e.recordAppliedSpec` which is not generic.